src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/GoalEntityMapper.kt

```diff
@@ -2,29 +2,90 @@ package uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.ma
 
 import org.mapstruct.Mapper
 import org.mapstruct.Mapping
+import org.mapstruct.MappingTarget
+import org.springframework.beans.factory.annotation.Autowired
 import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.entity.GoalEntity
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.entity.StepEntity
 import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.goal.Goal
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.goal.Step
 
 @Mapper(
   uses = [
     StepEntityMapper::class,
   ],
 )
-interface GoalEntityMapper {
+abstract class GoalEntityMapper {
+
+  @Autowired
+  private lateinit var stepEntityMapper: StepEntityMapper
 
   /**
    * Maps the supplied [Goal] into a new un-persisted [GoalEntity].
    * The JPA managed fields are not mapped.
    * This method is suitable for creating a new [GoalEntity] to be subsequently persisted to the database.
    */
   @ExcludeJpaManagedFieldsIncludingDisplayNameFields
-  fun fromDomainToEntity(goal: Goal): GoalEntity
+  abstract fun fromDomainToEntity(goal: Goal): GoalEntity
 
   /**
    * Maps the supplied [GoalEntity] into the domain [Goal].
    */
   @Mapping(target = "lastUpdatedBy", source = "updatedBy")
   @Mapping(target = "lastUpdatedByDisplayName", source = "updatedByDisplayName")
   @Mapping(target = "lastUpdatedAt", source = "updatedAt")
-  fun fromEntityToDomain(goalEntity: GoalEntity): Goal
+  abstract fun fromEntityToDomain(goalEntity: GoalEntity): Goal
+
+  /**
+   * Updates the supplied [GoalEntity] with fields from the supplied [Goal]. The updated [GoalEntity] can then be
+   * persisted to the database.
+   */
```

Comment:
The annotations tell MapStruct what fields to map, and how to do the mapping
There are certain fields we need to exclude from the mapping, such as the JPA fields (JPA is the Java Persistence API (ie. database), so we need to exclude fields like database ids etc, otherwise the database layer will get very upset when we try to save it to the database)
We also want to exclude any changes to the "reference" field, because that is the business key/id of the `Goal` (its not the DB primary key / ID; but we still dont want users changing it)